core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/DefaultIItemResultSet.java

```diff
@@ -89,29 +102,69 @@ private void calculateData(IItemCollection input) {
 								.getAggregate(aggregators.get(j));
 					}
-					data.add(row);
+					processingQueue.add(row);
 				}
 			}
 		} else {
 			IAggregator<?, ?> aggregator = Aggregators.distinct(query.getGroupBy());
+			Future future = null;
+			final IItemCollection newInput = input;
 			Set<?> aggregate = input.getAggregate((IAggregator<Set<?>, ?>) aggregator);
 			if (aggregate != null) {
-				for (Object o : aggregate) {
-					IItemCollection rowCollection = input.apply(ItemFilters.equals((IAttribute) query.getGroupBy(), o));
-					Object[] row = newRow();
-					int column = 0;
-					for (; column < attributes.size(); column++) {
-						// Optimization - it is too expensive to do aggregation for these. You simply
-						// get first non-null matching attribute - we're only using this for the group by today.
-						row[column] = getFirstNonNull(rowCollection, attributes.get(column));
+				try {
+					for (final Object o : aggregate) {
+						future = exec.submit(new Runnable() {
+							@Override
+							public void run() {
+								IItemCollection rowCollection = newInput
+										.apply(ItemFilters.equals((IAttribute) query.getGroupBy(), o));
+								Object[] row = newRow();
+								int column = 0;
+								for (; column < attributes.size(); column++) {
+									// Optimization - it is too expensive to do aggregation for these. You simply
+									// get first non-null matching attribute - we're only using this for the group by today.
+									row[column] = getFirstNonNull(rowCollection, attributes.get(column));
+								}
+								for (int j = 0; j < aggregators.size(); j++) {
+									row[column + j] = rowCollection.getAggregate(aggregators.get(j));
+								}
+								processingQueue.add(row);
+							}
+
+						});
+						futures.add(future);
 					}
-					for (int j = 0; j < aggregators.size(); j++) {
-						row[column + j] = rowCollection.getAggregate(aggregators.get(j));
+				} finally {
+					exec.shutdown();
+					try {
+						if (!exec.awaitTermination(configuredTimeout, TimeUnit.MINUTES)) {
+							checkStatusOfAllTasks(futures);
+
+							exec.shutdownNow();
+							// Await time for termination is not required as we have already given enough time before force shutdown. 
+						}
+					} catch (InterruptedException ie) {
+						exec.shutdownNow();
+						Thread.currentThread().interrupt();
 					}
-					data.add(row);
 				}
 			}
 		}
+		data.addAll(processingQueue);
+	}
+
+	private void checkStatusOfAllTasks(List<Future> totalAssignedTasks) {
+		int completedTask = 0;
+		for (Future future : totalAssignedTasks) {
+			if (future.isDone())
+				completedTask++;
+		}
+
+		if (completedTask < totalAssignedTasks.size()) {
+			System.out.println(
+					"The results may be inaccurate as JMC is unable to process all the class entries to determine "
+							+ "the class leak results. Please increase the configured timeout in preferences to see the accurate results.  ");
```

Comment:
I believe this should be changed to use the jul logger. Other flighrecorder component uses that.

E.g.

```java
private final static Logger LOGGER = Logger.getLogger("org.openjdk.jmc.flighrecorder.rules.jdk.util");
```

Reply:
Hi @bric3 Thanks for taking out time and reviewing the PR. I have replaced sysout with Logger now.